transaction/LockManager.cpp

```diff
@@ -0,0 +1,217 @@
+}
+
+bool LockManager::releaseAllLocks(const transaction_id tid) {
+  const std::vector<ResourceId> resource_ids
+      = transaction_table_->getResourceIdList(tid);
+  const TransactionTableResult transaction_deleted
+      = transaction_table_->deleteTransaction(tid);
+  if (transaction_deleted
+          == TransactionTableResult::kTransactionDeleteError) {
+    LOG(FATAL)
+        << "In LockManager.releaseAllLocks: Transaction could not be deleted!";
+  }
+
+  lock_table_->latchExclusive();
+
+  for (const auto &resource_id : resource_ids) {
+    const LockTableResult lock_deleted = lock_table_->deleteLock(tid, resource_id);
+
+    LOG(INFO) << "Transaction "
+                 + std::to_string(tid)
+                 + " released lock:"
+                 + resource_id.toString()
+                 + "\n";
+
+    if (lock_deleted == LockTableResult::kDeleteError) {
+      LOG(FATAL) << "In LockManager.releaseAllLock lock could not be deleted "
+                    "from LockTable";
+    }
+  }
+
+  lock_table_->unlatchExclusive();
+  return true;
+}
+
+bool LockManager::acquireLockInternal(const transaction_id tid,
+                                      const ResourceId &rid,
+                                      const AccessMode access_mode) {
+  const LockTableResult lock_result = lock_table_->putLock(tid, rid, access_mode);
+  CHECK(lock_result != LockTableResult::kPutError)
+      << "Unexpected result in LockManager.acquireLockInternal";
+
+  if (lock_result == LockTableResult::kAlreadyInOwned) {
+    return true;
+  } else if (lock_result == LockTableResult::kPlacedInOwned) {
+    const TransactionTableResult transaction_result
+        = transaction_table_->putOwnEntry(tid, rid, access_mode);
+    CHECK(transaction_result == TransactionTableResult::kPlacedInOwned)
+        << "Unexpected result in LockManager.acquireLockInternal: "
+           "Mismatch of table results (No.1).";
+    return true;
+  } else if (lock_result == LockTableResult::kAlreadyInPending) {
+    return false;
+  } else if (lock_result == LockTableResult::kPlacedInPending) {
+    const TransactionTableResult transaction_result =
+      transaction_table_->putPendingEntry(tid, rid, access_mode);
+    CHECK(transaction_result == TransactionTableResult::kPlacedInPending)
+        << "Unexpected result in LockManager.acquireLockInternal: "
+           "Mismatch of table results (No.2).";
+    return false;
+  }
+  return false;
+}
+
+void LockManager::killVictims() {
+  // TODO(Hakan): Find a method to latch this function
+  //              (it cannot because it calls releaseLocks)
+  if (detector_status_.load() == DeadLockDetectorStatus::kDone) {
+    for (const auto victim_transaction_id : victim_result_) {
+      releaseAllLocks(victim_transaction_id);
+      // TODO(Hakan): Find a way to kill transaction, so that requests with this
+      //              tid should be ignored.
+      LOG(INFO) << "Killed transaction " + std::to_string(victim_transaction_id);
```

Comment:
We may want to use `<<` instead of `+` before `std::string` to avoid temp string objects.